File: src/main/java/com/joecollins/graphics/screens/generic/AllSeatsScreen.java
package com.joecollins.graphics.screens.generic;

import static com.joecollins.models.general.PartyResult.NO_RESULT;

import com.joecollins.bindings.Bindable;
import com.joecollins.bindings.BindableList;
import com.joecollins.bindings.Binding;
import com.joecollins.bindings.BindingReceiver;
import com.joecollins.bindings.IndexedBinding;
import com.joecollins.graphics.components.ResultListingFrame;
import com.joecollins.graphics.utils.StandardFont;
import com.joecollins.models.general.Party;
import com.joecollins.models.general.PartyResult;
import java.awt.BorderLayout;
import java.awt.Color;
import java.awt.GridLayout;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.function.Function;
import java.util.stream.Collectors;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.border.EmptyBorder;
import org.apache.commons.lang3.tuple.ImmutablePair;
import org.apache.commons.lang3.tuple.ImmutableTriple;
import org.apache.commons.lang3.tuple.Pair;

public class AllSeatsScreen extends JPanel {

  private AllSeatsScreen(JLabel title, ResultListingFrame frame) {
    setBackground(Color.WHITE);
    setLayout(new BorderLayout());
    add(title, BorderLayout.NORTH);

    JPanel panel = new JPanel();
    panel.setBackground(Color.WHITE);
    panel.setBorder(new EmptyBorder(5, 5, 5, 5));
    panel.setLayout(new GridLayout(1, 1));
    panel.add(frame);
    add(panel, BorderLayout.CENTER);
  }

  public static <T> Builder<T> of(
      Binding<Map<T, Map<Party, Integer>>> prevResultBinding,
      Binding<Map<T, PartyResult>> currResultBinding,
      Function<T, String> nameFunc,
      Binding<String> headerBinding) {
    return new Builder<>(prevResultBinding, currResultBinding, nameFunc, headerBinding);
  }

  public static class Builder<T> {

    private final BindingReceiver<Map<T, Map<Party, Integer>>> prevResults;
    private final BindingReceiver<Map<T, PartyResult>> currResults;
    private final BindingReceiver<String> header;
    private final Function<T, String> nameFunc;

    private BindingReceiver<Integer> numRows = new BindingReceiver<>(() -> 20);
    private BindingReceiver<Set<T>> seatFilter = new BindingReceiver<>(Binding.fixedBinding(null));

    public Builder(
        Binding<Map<T, Map<Party, Integer>>> prevResultBinding,
        Binding<Map<T, PartyResult>> currResultBinding,
        Function<T, String> nameFunc,
        Binding<String> headerBinding) {
      this.prevResults = new BindingReceiver<>(prevResultBinding);
      this.currResults = new BindingReceiver<>(currResultBinding);
      this.header = new BindingReceiver<>(headerBinding);
      this.nameFunc = nameFunc;
    }

    public Builder<T> withNumRows(Binding<Integer> numRowsBinding) {
      this.numRows = new BindingReceiver<>(numRowsBinding);
      return this;
    }

    public Builder<T> withSeatFilter(Binding<Set<T>> seatFilterBinding) {
      this.seatFilter = new BindingReceiver<>(seatFilterBinding);
      return this;
    }

    public AllSeatsScreen build(Binding<String> titleBinding) {
      JLabel headerLabel = new JLabel();
      headerLabel.setFont(StandardFont.readBoldFont(32));
      headerLabel.setHorizontalAlignment(JLabel.CENTER);
      headerLabel.setBorder(new EmptyBorder(5, 0, -5, 0));
      titleBinding.bind(headerLabel::setText);

      Input<T> inputs = new Input<>(nameFunc);
      prevResults.getBinding().bind(inputs::setPrevResults);
      currResults.getBinding().bind(inputs::setCurrResults);
      seatFilter.getBinding().bind(inputs::setSeatFilter);

      BindableList<Entry<T>> entries = new BindableList<>();
      inputs.getResultBinding().bind(entries::setAll);

      ResultListingFrame frame = new ResultListingFrame();
      frame.setHeaderBinding(header.getBinding());
      frame.setNumRowsBinding(numRows.getBinding());
      frame.setNumItemsBinding(Binding.sizeBinding(entries));
      frame.setTextBinding(IndexedBinding.propertyBinding(entries, e -> nameFunc.apply(e.key)));
      frame.setBorderBinding(IndexedBinding.propertyBinding(entries, e -> e.prevColor));
      frame.setBackgroundBinding(
          IndexedBinding.propertyBinding(entries, e -> e.fill ? e.resultColor : Color.WHITE));
      frame.setForegroundBinding(
          IndexedBinding.propertyBinding(entries, e -> !e.fill ? e.resultColor : Color.WHITE));
      return new AllSeatsScreen(headerLabel, frame);
    }
  }

  private static class Input<T> extends Bindable<Input.Property> {
    private enum Property {
      PREV,
      CURR,
      FILTER
    }

    private final Function<T, String> nameFunc;
    private List<Pair<T, Party>> prevResults = List.of();
    private Map<T, PartyResult> currResults = Map.of();
    private Set<T> seatFilter = null;

    private Input(Function<T, String> nameFunc) {
      this.nameFunc = nameFunc;
    }

    public void setPrevResults(Map<T, Map<Party, Integer>> prevResults) {
      this.prevResults =
          prevResults.entrySet().stream()
              .map(
                  e -> {
                    Map<Party, Integer> votes = e.getValue();
                    int total = votes.values().stream().mapToInt(i -> i).sum();
                    List<Integer> topTwo =
                        votes.values().stream()
                            .sorted(Comparator.reverseOrder())
                            .limit(2)
                            .collect(Collectors.toList());
                    return ImmutablePair.of(e, 1.0 * (topTwo.get(0) - topTwo.get(1)) / total);
                  })
              .sorted(Comparator.comparing(e -> nameFunc.apply(e.getLeft().getKey()).toUpperCase()))
              .map(Map.Entry::getKey)
              .map(
                  e ->
                      ImmutablePair.of(
                          e.getKey(),
                          e.getValue().entrySet().stream()
                              .max(Map.Entry.comparingByValue())
                              .map(Map.Entry::getKey)
                              .orElse(null)))
              .collect(Collectors.toList());
      onPropertyRefreshed(Input.Property.PREV);
    }

    public void setCurrResults(Map<T, PartyResult> currResults) {
      this.currResults = currResults;
      onPropertyRefreshed(Input.Property.CURR);
    }

    public void setSeatFilter(Set<T> seatFilter) {
      this.seatFilter = seatFilter;
      onPropertyRefreshed(Input.Property.FILTER);
    }

    public Binding<List<Entry<T>>> getResultBinding() {
      return Binding.propertyBinding(
          this,
          t ->
              t.prevResults.stream()
                  .filter(e -> t.seatFilter == null || t.seatFilter.contains(e.getKey()))
                  .map(
                      e ->
                          ImmutableTriple.of(
                              e.getLeft(),
                              e.getRight(),
                              t.currResults.getOrDefault(e.getLeft(), NO_RESULT)))
                  .map(
                      e -> {
                        var result = e.getRight() == null ? NO_RESULT : e.getRight();
                        return new Entry<>(
                            e.getLeft(),
                            e.getMiddle().getColor(),
                            result.getParty() == null ? Color.BLACK : result.getParty().getColor(),
                            result.isElected());
                      })
                  .collect(Collectors.toList()),
          Input.Property.PREV,
          Input.Property.CURR,
          Input.Property.FILTER);
    }
  }

  private static class Entry<T> {
    private final T key;
    private final Color prevColor;
    private final Color resultColor;
    private final boolean fill;

    private Entry(T key, Color prevColor, Color resultColor, boolean fill) {
      this.key = key;
      this.prevColor = prevColor;
      this.resultColor = resultColor;
      this.fill = fill;
    }
  }
}

File: src/main/java/com/joecollins/graphics/screens/generic/AllSeatsScreen.kt
package com.joecollins.graphics.screens.generic

import com.joecollins.bindings.Bindable
import com.joecollins.bindings.Binding
import com.joecollins.bindings.BindingReceiver
import com.joecollins.bindings.mapElements
import com.joecollins.graphics.components.FontSizeAdjustingLabel
import com.joecollins.graphics.components.ResultListingFrame
import com.joecollins.graphics.utils.StandardFont.readBoldFont
import com.joecollins.models.general.Party
import com.joecollins.models.general.PartyResult
import com.joecollins.pubsub.Publisher
import com.joecollins.pubsub.asOneTimePublisher
import com.joecollins.pubsub.mapElements
import java.awt.BorderLayout
import java.awt.Color
import java.awt.GridLayout
import java.util.concurrent.Flow
import javax.swing.JLabel
import javax.swing.JPanel
import javax.swing.border.EmptyBorder

class AllSeatsScreen private constructor(title: JLabel, frame: ResultListingFrame) : JPanel() {
    class Builder<T>(
        prevResultBinding: Binding<out Map<T, Map<Party, Int>>>,
        currResultBinding: Binding<out Map<T, PartyResult?>>,
        private val nameFunc: (T) -> String,
        headerPublisher: Flow.Publisher<out String?>
    ) {
        private val prevResults: BindingReceiver<Map<T, Map<Party, Int>>> = BindingReceiver(prevResultBinding)
        private val currResults: BindingReceiver<Map<T, PartyResult?>> = BindingReceiver(currResultBinding)
        private val header: Flow.Publisher<out String?> = headerPublisher
        private var numRows: Flow.Publisher<out Int> = 20.asOneTimePublisher()
        private var seatFilter = BindingReceiver<Set<T>?>(Binding.fixedBinding(null))

        fun withNumRows(numRowsPublisher: Flow.Publisher<out Int>): Builder<T> {
            numRows = numRowsPublisher
            return this
        }

        fun withSeatFilter(seatFilterBinding: Binding<out Set<T>?>): Builder<T> {
            seatFilter = BindingReceiver(seatFilterBinding)
            return this
        }

        fun build(titleBinding: Binding<out String?>): AllSeatsScreen {
            val headerLabel = FontSizeAdjustingLabel()
            headerLabel.font = readBoldFont(32)
            headerLabel.horizontalAlignment = JLabel.CENTER
            headerLabel.border = EmptyBorder(5, 0, -5, 0)
            titleBinding.bind { headerLabel.text = it }
            val inputs = Input(nameFunc)
            prevResults.getBinding().bind { inputs.setPrevResults(it) }
            currResults.getBinding().bind { inputs.setCurrResults(it) }
            seatFilter.getBinding().bind { inputs.setSeatFilter(it) }
            val frame = ResultListingFrame(
                headerPublisher = header,
                numRowsPublisher = numRows,
                itemsPublisher = inputs.resultPublisher.mapElements {
                    ResultListingFrame.Item(
                        text = nameFunc(it.key),
                        border = it.prevColor,
                        background = if (it.fill) it.resultColor else Color.WHITE,
                        foreground = if (!it.fill) it.resultColor else Color.WHITE
                    )
                }
            )
            return AllSeatsScreen(headerLabel, frame)
        }
    }

    private class Input<T>(private val nameFunc: (T) -> String) : Bindable<Input<T>, Input.Property>() {
        private enum class Property {
            PREV, CURR, FILTER
        }

        private var prevResults: List<Pair<T, Party>> = emptyList()
        private var currResults: Map<T, PartyResult?> = emptyMap()
        private var seatFilter: Set<T>? = null

        fun setPrevResults(prevResults: Map<T, Map<Party, Int>>) {
            this.prevResults = prevResults.entries
                .asSequence()
                .map { e ->
                    val votes = e.value
                    val total = votes.values.sum()
                    val topTwo = votes.values
                        .sortedDescending()
                        .take(2)
                        .toList()
                    Pair(e, 1.0 * (topTwo[0] - topTwo[1]) / total)
                }
                .sortedBy { e -> nameFunc(e.first.key).uppercase() }
                .map { it.first }
                .map { e ->
                    Pair(
                        e.key,
                        e.value.entries
                            .maxByOrNull { it.value }
                        !!.key
                    )
                }
                .toList()
            _resultPublisher.submit(toEntries())
        }

        fun setCurrResults(currResults: Map<T, PartyResult?>) {
            this.currResults = currResults
            _resultPublisher.submit(toEntries())
        }

        fun setSeatFilter(seatFilter: Set<T>?) {
            this.seatFilter = seatFilter
            _resultPublisher.submit(toEntries())
        }

        val _resultPublisher = Publisher(toEntries())
        val resultPublisher: Flow.Publisher<out List<Entry<T>>> = _resultPublisher

        private fun toEntries() = this.prevResults
            .asSequence()
            .filter { e: Pair<T, Party> -> this.seatFilter?.contains(e.first) ?: true }
            .map { e: Pair<T, Party> ->
                Triple(
                    e.first,
                    e.second,
                    this.currResults[e.first] ?: PartyResult.NO_RESULT
                )
            }
            .map { e: Triple<T, Party, PartyResult?> ->
                val result = e.third ?: PartyResult.NO_RESULT
                Entry(
                    e.first,
                    e.second.color,
                    result.party?.color ?: Color.BLACK,
                    result.isElected
                )
            }
            .toList()
    }

    private class Entry<T>(val key: T, val prevColor: Color, val resultColor: Color, val fill: Boolean)
    companion object {
        @JvmStatic fun <T> of(
            prevResultBinding: Binding<out Map<T, Map<Party, Int>>>,
            currResultBinding: Binding<out Map<T, PartyResult?>>,
            nameFunc: (T) -> String,
            headerPublisher: Flow.Publisher<out String?>
        ): Builder<T> {
            return Builder(prevResultBinding, currResultBinding, nameFunc, headerPublisher)
        }
    }

    init {
        background = Color.WHITE
        layout = BorderLayout()
        add(title, BorderLayout.NORTH)
        val panel = JPanel()
        panel.background = Color.WHITE
        panel.border = EmptyBorder(5, 5, 5, 5)
        panel.layout = GridLayout(1, 1)
        panel.add(frame)
        add(panel, BorderLayout.CENTER)
    }
}
